// @fal-ai/serverless-client/src/function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.queue = exports.subscribe = exports.run = exports.buildUrl = void 0;
const tslib_1 = require("tslib");
const config_1 = require("./config");
const request_1 = require("./request");
const storage_1 = require("./storage");
const utils_1 = require("./utils");
/**
 * Builds the final url to run the function based on its `id` or alias and
 * a the options from `RunOptions<Input>`.
 *
 * @private
 * @param id the function id or alias
 * @param options the run options
 * @returns the final url to run the function
 */
function buildUrl(id, options = {}) {
    var _a, _b;
    const { host } = (0, config_1.getConfig)();
    const method = ((_a = options.method) !== null && _a !== void 0 ? _a : 'post').toLowerCase();
    const path = ((_b = options.path) !== null && _b !== void 0 ? _b : '').replace(/^\//, '').replace(/\/{2,}/, '/');
    const input = options.input;
    const params = 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    method === 'get' && input ? new URLSearchParams(input) : undefined;
    const queryParams = params ? `?${params.toString()}` : '';
    const parts = id.split('/');
    // if a fal.ai url is passed, just use it
    if ((0, utils_1.isValidUrl)(id)) {
        const url = id.endsWith('/') ? id : `${id}/`;
        return `${url}${path}${queryParams}`;
    }
    if (parts.length === 2 && (0, utils_1.isUUIDv4)(parts[1])) {
        return `https://${host}/trigger/${id}/${path}${queryParams}`;
    }
    return `https://${id}.${host}/${path}${queryParams}`;
}
exports.buildUrl = buildUrl;
/**
 * Runs a fal serverless function identified by its `id`.
 *
 * @param id the registered function revision id or alias.
 * @returns the remote function output
 */
function run(id, options = {}) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const input = options.input && options.autoUpload !== false
            ? yield storage_1.storageImpl.transformInput(options.input)
            : options.input;
        return (0, request_1.dispatchRequest)((_a = options.method) !== null && _a !== void 0 ? _a : 'post', buildUrl(id, options), input);
    });
}
exports.run = run;
/**
 * Subscribes to updates for a specific request in the queue.
 *
 * @param id - The ID or URL of the function web endpoint.
 * @param options - Options to configure how the request is run and how updates are received.
 * @returns A promise that resolves to the result of the request once it's completed.
 */
function subscribe(id, options = {}) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { request_id: requestId } = yield exports.queue.submit(id, options);
        if (options.onEnqueue) {
            options.onEnqueue(requestId);
        }
        return new Promise((resolve, reject) => {
            var _a;
            let timeoutId;
            const pollInterval = (_a = options.pollInterval) !== null && _a !== void 0 ? _a : 1000;
            const poll = () => tslib_1.__awaiter(this, void 0, void 0, function* () {
                var _b;
                try {
                    const requestStatus = yield exports.queue.status(id, {
                        requestId,
                        logs: (_b = options.logs) !== null && _b !== void 0 ? _b : false,
                    });
                    if (options.onQueueUpdate) {
                        options.onQueueUpdate(requestStatus);
                    }
                    if (requestStatus.status === 'COMPLETED') {
                        clearTimeout(timeoutId);
                        try {
                            const result = yield exports.queue.result(id, { requestId });
                            resolve(result);
                        }
                        catch (error) {
                            reject(error);
                        }
                        return;
                    }
                    timeoutId = setTimeout(poll, pollInterval);
                }
                catch (error) {
                    clearTimeout(timeoutId);
                    reject(error);
                }
            });
            poll().catch(reject);
        });
    });
}
exports.subscribe = subscribe;
/**
 * The fal run queue module. It allows to submit a function to the queue and get its result
 * on a separate call. This is useful for long running functions that can be executed
 * asynchronously and not .
 */
exports.queue = {
    submit(id, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { webhookUrl, path = '' } = options, runOptions = tslib_1.__rest(options, ["webhookUrl", "path"]);
            const query = webhookUrl
                ? '?' + new URLSearchParams({ fal_webhook: webhookUrl }).toString()
                : '';
            return run(id, Object.assign(Object.assign({}, runOptions), { method: 'post', path: '/fal/queue/submit' + path + query }));
        });
    },
    status(id, { requestId, logs = false }) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return run(id, {
                method: 'get',
                path: `/fal/queue/requests/${requestId}/status`,
                input: {
                    logs: logs ? '1' : '0',
                },
            });
        });
    },
    result(id, { requestId }) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return run(id, {
                method: 'get',
                path: `/fal/queue/requests/${requestId}/response`,
            });
        });
    },
    subscribe,
};
